design/legacy-design/device-service/discovery.md

```diff
@@ -0,0 +1,79 @@
+
+A boolean configuration value `Device/Discovery` defaults to false. If this
+value is set true, and the DS implementation supports discovery, discovery is
+enabled.
+
+The SDK will respond to POST requests on the the /discovery endpoint. This call
+will return one of the following codes:
+
+* 200: discovery has been triggered or is already running
+* 423: the service is locked (admin state) or disabled (operating state)
+* 500: unknown or unanticipated issues exist
+* 501: discovery is not supported in this implementation
+* 503: discovery is disabled by configuration
+
+In the case where discovery is triggered, the discovery process will run in a
+new thread or goroutine, so that the REST call may return immediately.
+
+An integer configuration value `Device/DiscoveryInterval` defaults to zero. If
+this value is set to a positive value, and discovery is enabled, the discovery
+process will be triggered at the specified interval (in seconds).
+
+### Finding Devices
+
+When discovery is triggered, the SDK calls the implementation function provided
+by the Device Service. This should perform whatever protocol-specific procedure
+is necessary to find devices, and for each device found, call the filtered
+device addition function provided by the SDK.
+
+The information required for a found device is as follows:
+
+* An autogenerated device name
+* The Protocol Properties of the device
+* Optionally, a description string
+* Optionally, a list of label strings
+
+**Rationale:** *An alternative design would have the implementation function return
+a collection of discovered devices to the SDK. Using a callback mechanism instead has the following advantages:*
+
+* *Avoids collection processing in languages without builtin support for such, eg C*
+* *Allows for asynchronous operation, where the DS implementation initiates discovery and returns immediately, then when results come back (eg, from a network), handler functions in the implementation call the filtered device addition function themselves*
+* *Allows DS implementations where devices self-announce to call the filtered device addition function independent of the discovery process*
+
+
+### Filtered Device Addition
+
+The filter criteria for discovered devices are represented by Provision Watchers. A Provision Watcher contains the following fields:
+
+* `Identifiers`: A set of name-value pairs against which a new device's ProtocolProperties are matched
+* `BlockingIdentifiers`: A further set of name-value pairs which are also matched against a new device's ProtocolProperties
+* `Profile`: The name of a DeviceProfile which should be assigned to new devices which pass this ProvisionWatcher
+* `AdminState`: The initial Administrative State for new devices which pass this ProvisionWatcher
+
+A candidate new device passes a ProvisionWatcher if all of the `Identifiers` match, and none of the `BlockingIdentifiers`.
+
+The values specified in `Identifiers` may be regular expressions.
```

Comment:
How does the SDK know if they're regular expressions or not? Why not just stipulate that they _are_ regular expressions?

Reply:
Yes. Changed this.